### PR TITLE
Migrate TPU tests to GitHub actions

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Replaces CircleCI for GitHub actions to launch TPU tests.
This has the benefit that we can make TPU tests work on forks and mark them as required again.

Closes #14307
Part of https://github.com/Lightning-AI/lightning/issues/13572

### Does your PR introduce any breaking changes? If yes, please list them.

None

cc @carmocca @akihironitta @borda

### CODE REVIEW

File: .github/workflows/ci-pytorch-test-tpu.yml [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+       data = data.replace('{PYTHON_VERSION}', '$PYTHON_VER').replace('{PR_NUMBER}', '$PR_NUMBER').replace('{SHA}', '$SHA') ; open(fname, 'w').write(data)"
+       cat dockers/tpu-tests/tpu_test_cases.jsonnet
+      shell: bash
+
+    # https://docs.github.com/en/actions/deployment/deploying-to-your-cloud-provider/deploying-to-google-kubernetes-engine
+    - uses: google-github-actions/get-gke-credentials@fb08709ba27618c31c09e014e1d8364b02e5042e
+      with:
+        cluster_name: TODO
+        location: us-central1-c
+        credentials: ${{ secrets.GKE_SA_KEY }}
```
Comment (carmocca) [inline .github/workflows/ci-pytorch-test-tpu.yml:72]: No idea if this is correct.

**Comment by carmocca** [comment]
This is working!! :open_mouth:

**Comment by carmocca** [comment]
Waiting for https://github.com/Lightning-AI/lightning/pull/14926

File: .github/workflows/events-recurrent.yml [hunk 2]
```diff
@@ -1,43 +0,0 @@
...
-  schedule:
-    - cron: "*/20 * * * *" # At every 20 minutes
-
-env:
-  GKE_CLUSTER: lightning-cluster
-  GKE_ZONE: us-central1-a
-
-jobs:
-  tpu-cleanup:
-    name: TPU cleaning
```
Comment (carmocca) [inline .github/workflows/events-recurrent.yml:17]: I don't understand why we had this if the circleci and kubernetes job had timeouts

File: .github/workflows/ci-pytorch-test-tpu.yml [hunk 3]
```diff
@@ -0,0 +1,131 @@
+on:
+  push:
+    branches: [master, "release/*"]
+    paths:
+      - ".github/workflows/ci-pytorch-test-tpu.yml"
+      - "requirements/pytorch/**"
+      - "src/pytorch_lightning/**"
+      - "tests/tests_pytorch/**"
+      - "setup.cfg"  # includes pytest config
+  # FIXME: use _target after merge to share secrets
```
Comment (carmocca) [inline .github/workflows/ci-pytorch-test-tpu.yml:10]:
```suggestion
  # TODO: use _target after merge to share secrets
```

File: .github/workflows/ci-pytorch-test-tpu.yml [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+env:
+  PROJECT_ID: ${{ secrets.GKE_PROJECT }}
+  GKE_CLUSTER: lightning-cluster
+  GKE_ZONE: us-central1-a
+
+jobs:
+  test-on-tpus:
+    runs-on: ubuntu-22.04
+    # FIXME: uncomment after finished
+    # if: github.event.pull_request.draft == false
```
Comment (carmocca) [inline .github/workflows/ci-pytorch-test-tpu.yml:34]:
```suggestion
    if: github.event.pull_request.draft == false
```

**Review by awaelchli — APPROVED** [review]

File: .github/workflows/ci-pytorch-test-tpu.yml [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+       done
+       echo "Done waiting. Job status code: $status_code"
+       kubectl logs -f $pod_name --container=train > /tmp/full_output.txt
+       if grep -q '<?xml version="1.0" ?>' /tmp/full_output.txt; then
+         csplit /tmp/full_output.txt '/<?xml version="1.0" ?>/';
+       else
+         mv /tmp/full_output.txt xx00;
+       fi
+       # First portion is the test logs.
+       cat xx00 && echo "Done with log retrieval attempt."
```
Comment (carmocca) [inline .github/workflows/ci-pytorch-test-tpu.yml:109]: This `xx00` file is empty so `coverage report` fails below with `No data to report.`

Any ideas @akihironitta @Borda?

**Review by otaj — APPROVED** [review]

File: .github/workflows/ci-pytorch-test-tpu.yml (hunk 5 above)
Comment (Borda) [reply .github/workflows/ci-pytorch-test-tpu.yml:109]: How about the `/tmp/full_output.txt` above? Is it also empty?
cc: @kaushikb11

File: .github/workflows/ci-pytorch-test-tpu.yml [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+         elif kubectl get jobs $job_name -o jsonpath='Succeeded:{.status.succeeded}' | grep "Succeeded:1"; then
+           status_code=0 && break;
+         else
+           printf ".";
+         fi;
+         sleep 5;
+       done
+       echo "Done waiting. Job status code: $status_code"
+       kubectl logs -f $pod_name --container=train > /tmp/full_output.txt
+       cat /tmp/full_output.txt
```
Comment (carmocca) [inline .github/workflows/ci-pytorch-test-tpu.yml:103]: FIXME

**Review by Borda — APPROVED** [review]
asked @lexierule to setup secrets :otter:

**Comment by carmocca** [comment]
@Borda the secrets are already set up and working. This is just blocked by coverage issues

**Comment by Borda** [comment]
Secrets shall be set, thx @lantiga :rabbit:

**Pull request merged**